apps/emqx/src/bpapi/emqx_bpapi_static_checks.erl

```diff
@@ -0,0 +1,263 @@
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_bpapi_static_checks).
+
+-export([dump/1, dump/0, check_compat/1]).
+
+-include_lib("emqx/include/logger.hrl").
+
+-type api_dump() :: #{{emqx_bpapi:api(), emqx_bpapi:api_version()} =>
+                          #{ calls := [emqx_bpapi:rpc()]
+                           , casts := [emqx_bpapi:rpc()]
+                           }}.
+
+-type dialyzer_spec() :: {_Type, [_Type]}.
+
+-type dialyzer_dump() :: #{mfa() => dialyzer_spec()}.
+
+-type fulldump() :: #{ api        => api_dump()
+                     , signatures => dialyzer_dump()
+                     , release    => string()
+                     }.
+
+-type param_types() :: #{emqx_bpapi:var_name() => _Type}.
+
+%% Applications we wish to ignore in the analysis:
+-define(IGNORED_APPS, "gen_rpc, recon, observer_cli, snabbkaffe, ekka, mria").
+%% List of known RPC backend modules:
+-define(RPC_MODULES, "gen_rpc, erpc, rpc, emqx_rpc").
+%% List of functions in the RPC backend modules that we can ignore:
+-define(IGNORED_RPC_CALLS, "gen_rpc:nodes/0").
+
+-define(XREF, myxref).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Functions related to BPAPI compatibility checking
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec check_compat([file:filename()]) -> boolean().
+check_compat(DumpFilenames) ->
+    put(bpapi_ok, true),
+    Dumps = lists:map(fun(FN) ->
+                              {ok, [Dump]} = file:consult(FN),
+                              Dump
+                      end,
+                      DumpFilenames),
+    [check_compat(I, J) || I <- Dumps, J <- Dumps],
+    erase(bpapi_ok).
+
+%% Note: sets nok flag
+-spec check_compat(fulldump(), fulldump()) -> ok.
+check_compat(Dump1, Dump2) ->
+    check_api_immutability(Dump1, Dump2),
+    typecheck_apis(Dump1, Dump2).
+
+%% It's not allowed to change BPAPI modules. Check that no changes
+%% have been made. (sets nok flag)
+-spec check_api_immutability(fulldump(), fulldump()) -> ok.
+check_api_immutability(#{release := Rel1, api := APIs1}, #{release := Rel2, api := APIs2})
+  when Rel2 >= Rel1 ->
+    %% TODO: Handle API deprecation
+    _ = maps:map(
+          fun(Key = {API, Version}, Val) ->
+                  case maps:get(Key, APIs2, undefined) of
+                      Val ->
+                          ok;
+                      undefined ->
+                          setnok(),
+                          ?ERROR("API ~p v~p was removed in release ~p without being deprecated.",
+                                 [API, Version, Rel2]);
+                      _Val ->
```

Comment:
Q: since the arguments are the var names extracted, couldn't this give a false positive if the variable names are just refactored to other names?

Reply:
In theory renaming could be handled, but it goes against the assumption that backplane APIs are never changed.